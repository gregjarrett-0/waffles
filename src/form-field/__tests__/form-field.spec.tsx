import React, { useRef, useEffect } from 'react';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/react';

import { FormField } from '../index';

const sizes = ['small', 'medium', 'large'] as const;
const MOCKED_ID = '123abC';

jest.mock('nanoid', () => {
  return {
    nanoid: () => MOCKED_ID,
  };
});

function TestRefFormField() {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <FormField label="Focused with ref">
      <input ref={inputRef} />
    </FormField>
  );
}

type TestInputProps = {
  error?: boolean;
  inverted?: boolean;
} & React.InputHTMLAttributes<HTMLInputElement>;

function TestInput({
  error = false,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  inverted = false,
  ...props
}: TestInputProps) {
  return <input {...props} aria-invalid={error} />;
}

describe('FormField', () => {
  it('enhances basic input with label and description', () => {
    const { container, getByText } = render(
      <FormField label="Form field label" description="Form field description">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const label = container.querySelector('label');
    const description = getByText('Form field description');

    expect(input).toBeInTheDocument();
    expect(label).toHaveTextContent('Form field label');
    expect(description).toBeInTheDocument();
  });

  it('after clicking on label input got focused', () => {
    const handleFocus = jest.fn();
    const { getByText } = render(
      <FormField label="How do you feel?" description="Additional description">
        <TestInput onFocus={handleFocus} />
      </FormField>,
    );

    const label = getByText('How do you feel?');
    userEvent.click(label);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('form field child component and label are associated by the same ID', () => {
    const { container } = render(
      <FormField label="Form field label">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const label = container.querySelector('label');

    expect(label).toHaveAttribute('for', `form-field-${MOCKED_ID}`);
    expect(input).toHaveAttribute('id', `form-field-${MOCKED_ID}`);
  });

  it('child input component accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefFormField />);

    const input = container.querySelector('input');

    expect(input).toHaveFocus();
  });

  it('allow props such as required to be passed directly to the child component', () => {
    const { container } = render(
      <FormField label="Form field label">
        <TestInput required />
      </FormField>,
    );

    const input = container.querySelector('input');

    expect(input).toHaveAttribute('required');
  });

  it('allow props such as required, and disabled to be passed directly to the child component', () => {
    const { container } = render(
      <FormField label="Form field label">
        <TestInput required disabled />
      </FormField>,
    );

    const input = container.querySelector('input');

    expect(input).toHaveAttribute('required');
    expect(input).toBeDisabled();
  });

  it("when 'requiredIndicator' is set to 'showRequired', and required prop is passed display 'Required' indicator", () => {
    const { container, getByText } = render(
      <FormField
        label="Form field label"
        requiredIndicator="showRequired"
        required
      >
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const indicator = getByText(/required/i);

    expect(input).toHaveAttribute('required');
    expect(indicator).toBeInTheDocument();
  });

  it("when 'requiredIndicator' is set to 'showRequired', and required prop is not passed doesn't display indicator", () => {
    const { container, queryByText } = render(
      <FormField label="Form field label" requiredIndicator="showRequired">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const indicator = queryByText(/required/i);

    expect(input).not.toHaveAttribute('required');
    expect(indicator).not.toBeInTheDocument();
  });

  it("when 'requiredIndicator' is set to 'showOptional', and required prop is passed doesn't display indicator", () => {
    const { container, queryByText } = render(
      <FormField
        label="Form field label"
        requiredIndicator="showOptional"
        required
      >
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const indicator = queryByText(/optional/i);

    expect(input).toHaveAttribute('required');
    expect(indicator).not.toBeInTheDocument();
  });

  it("when 'requiredIndicator' is set to 'showOptional', and required prop is not passed display 'Optional' indicator", () => {
    const { container, getByText } = render(
      <FormField label="Form field label" requiredIndicator="showOptional">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const indicator = getByText(/optional/i);

    expect(input).not.toHaveAttribute('required');
    expect(indicator).toBeInTheDocument();
  });

  it("when 'requiredIndicator' is set to 'none', and required prop is not passed don't display any indicator", () => {
    const { container, queryByText } = render(
      <FormField label="Form field label" requiredIndicator="none">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const optional = queryByText(/optional/i);
    const required = queryByText(/required/i);

    expect(input).not.toHaveAttribute('required');
    expect(optional).not.toBeInTheDocument();
    expect(required).not.toBeInTheDocument();
  });

  it("when 'requiredIndicator' is set to 'none', and required prop is passed don't display any indicator", () => {
    const { container, queryByText } = render(
      <FormField label="Form field label" required requiredIndicator="none">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const optional = queryByText(/optional/i);
    const required = queryByText(/required/i);

    expect(input).toHaveAttribute('required');
    expect(optional).not.toBeInTheDocument();
    expect(required).not.toBeInTheDocument();
  });

  it('renders error message', () => {
    const { getByText } = render(
      <FormField label="Form field label" error="Form field error">
        <TestInput />
      </FormField>,
    );

    const error = getByText('Form field error');

    expect(error).toBeInTheDocument();
  });

  it('form field child component and error are associated by the same ID, and aria-invalid attribute is set', () => {
    const { container, getByText } = render(
      <FormField label="Form field label" error="Form field error">
        <TestInput />
      </FormField>,
    );

    const input = container.querySelector('input');
    const error = getByText('Form field error');

    expect(input).toHaveAttribute(
      'aria-errormessage',
      `form-field-error-${MOCKED_ID}`,
    );
    expect(input).toHaveAttribute('aria-invalid', 'true');
    expect(error).toHaveAttribute('id', `form-field-error-${MOCKED_ID}`);
  });

  it('renders snapshot of form field', () => {
    const { container } = render(
      <FormField
        label="Form field label"
        description="Form field description."
        error="Provided value is incorrect."
        required
        requiredIndicator="showRequired"
      >
        <TestInput />
      </FormField>,
    );

    const formField = container.firstChild;
    expect(formField).toMatchSnapshot();
  });

  it('renders snapshot of inverted form field', () => {
    const { container } = render(
      <FormField
        inverted
        label="Inverted form field label"
        description="Inverted form field description."
        error="Provided value is incorrect."
        required
        requiredIndicator="showRequired"
      >
        <TestInput />
      </FormField>,
    );

    const formField = container.firstChild;
    expect(formField).toMatchSnapshot();
  });
});

describe('renders snapshot of', () => {
  sizes.forEach((size) => {
    it(`size ${size}`, () => {
      const { container } = render(
        <FormField
          size={size}
          label="Form field label"
          description="Form field description"
        >
          <TestInput />
        </FormField>,
      );

      const formField = container.firstChild;
      expect(formField).toMatchSnapshot();
    });
  });
});
