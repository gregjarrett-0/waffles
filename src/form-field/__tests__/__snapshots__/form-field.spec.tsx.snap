// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders snapshot of form field 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-size: 14px;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  color: #5D6A77;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 6px;
  height: 6px;
  border-radius: 50%;
  background-color: #DD3400;
  margin-right: 4px;
}

.emotion-4 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  font-size: 14px;
  margin-top: -4px;
}

.emotion-5 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  margin-top: 8px;
  color: #C01100;
  font-size: 14px;
  line-height: 1.25;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-field-123abC"
  >
    Form field label
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      <span
        class="emotion-3"
      />
      Required
    </span>
  </label>
  <p
    class="emotion-4"
  >
    Form field description.
  </p>
  <input
    aria-describedby="form-field-error-123abC"
    aria-errormessage="form-field-error-123abC"
    aria-invalid="true"
    id="form-field-123abC"
    required=""
  />
  <p
    class="emotion-5"
    id="form-field-error-123abC"
  >
    Provided value is incorrect.
  </p>
</div>
`;

exports[`FormField renders snapshot of inverted form field 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-size: 14px;
  font-weight: 800;
  color: #FFFFFF;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  color: #9BA3AB;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 6px;
  height: 6px;
  border-radius: 50%;
  background-color: #FF5400;
  margin-right: 4px;
}

.emotion-4 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #FFFFFF;
  font-size: 14px;
  line-height: 1.25;
  font-size: 14px;
  margin-top: -4px;
}

.emotion-5 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  margin-top: 8px;
  color: #FF5400;
  font-size: 14px;
  line-height: 1.25;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-field-123abC"
  >
    Inverted form field label
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      <span
        class="emotion-3"
      />
      Required
    </span>
  </label>
  <p
    class="emotion-4"
  >
    Inverted form field description.
  </p>
  <input
    aria-describedby="form-field-error-123abC"
    aria-errormessage="form-field-error-123abC"
    aria-invalid="true"
    id="form-field-123abC"
    required=""
  />
  <p
    class="emotion-5"
    id="form-field-error-123abC"
  >
    Provided value is incorrect.
  </p>
</div>
`;

exports[`renders snapshot of size large 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-size: 14px;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  color: #5D6A77;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  font-size: 14px;
  margin-top: -4px;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-field-123abC"
  >
    Form field label
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      Optional
    </span>
  </label>
  <p
    class="emotion-3"
  >
    Form field description
  </p>
  <input
    aria-invalid="false"
    id="form-field-123abC"
  />
</div>
`;

exports[`renders snapshot of size medium 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-size: 14px;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  color: #5D6A77;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  font-size: 14px;
  margin-top: -4px;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-field-123abC"
  >
    Form field label
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      Optional
    </span>
  </label>
  <p
    class="emotion-3"
  >
    Form field description
  </p>
  <input
    aria-invalid="false"
    id="form-field-123abC"
  />
</div>
`;

exports[`renders snapshot of size small 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-size: 12px;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 10px;
  color: #5D6A77;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  font-size: 12px;
  margin-top: -4px;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-field-123abC"
  >
    Form field label
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      Optional
    </span>
  </label>
  <p
    class="emotion-3"
  >
    Form field description
  </p>
  <input
    aria-invalid="false"
    id="form-field-123abC"
  />
</div>
`;
